### PR TITLE
Add support for Mars Express HRSC SRC instrument

### PR DESCRIPTION
**Pull request by scsides**
Initial PR for adding support for the MEX HRSC SRC instrument to ISIS

## Description
Modify hrsc2isis to work with SRC files
Add serial number for SRC
Modify and add tests

## Motivation and Context
ISIS does not work with SRC image.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [X] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [X] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/mex/apps/hrsc2isis/MexHrscBandBin.trn [hunk 1]
```diff
@@ -1,32 +1,12 @@
...
-# the information.
-#
-# InputDefault is the value used if there is no value for
-# the keyword
-#
-# Translation is the native and corresponding foreign values.
-# Translation may be repeated as needed.
-#
+# Translations for HRSC Stereo and SRC instrument.
+# ISIS cube lable BandBen group
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/MexHrscBandBin.trn]: Single typo -- BandBen -> BandBin.

File: isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn [hunk 2]
```diff
@@ -1,25 +1,6 @@
...
-# the information.
-#
-# InputDefault is the value used if there is no value for
-# the keyword
-#
-# Translation is the native and corresponding foreign values.
-# Translation may be repeated as needed.
-#
+# Translations for HRSC Stereo and SRC instrument.
+# ISIS cube lable Instrument group
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn]: lable -> label

File: isis/src/mex/apps/hrsc2isis/hrsc2isis.xml [hunk 3]
```diff
@@ -5,18 +5,27 @@
...
+      table (LineScanTimes) with the prefix data and lineInFile with whether or 
+      not a gap should be inserted. For all files, we now process the data and if 
+      there were gaps, they will be put in their proper places. This is a two-pass
+      system for files with prefix data, one-pass for files without.  The Isis2
+      equivalent to this program is mex2isis.pl. It is worth noting that
+      regardless of the  input file's byte order, the prefix data byte order is 
+      always LSB.
+    </p>
+    <p>
+      For the SRC instrument,
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/hrsc2isis.xml]: Rest of content here is in progress?

File: isis/src/mex/apps/hrsc2isis/hrsc2isis.xml [hunk 4]
```diff
@@ -5,18 +5,27 @@
...
-    equivalent to this program is mex2isis.pl. It is worth noting that
-    regardless of the  input file's byte order, the prefix data byte order is 
-    always LSB.
+    <p>
+      This program imports Mars Express HRSC files. This includes both the HRSC 
+      Stereo and the SRC instruments with procesing level of "1" or "2" (not "3"). 
+    </p>
+    <p>
+      For the stereo instrument, we first
+      determining whether or not the input file has prefix data. If there is
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/hrsc2isis.xml]: determining -> determine

File: isis/src/mex/apps/hrsc2isis/tsts/srcImage/Makefile [hunk 5]
```diff
@@ -2,13 +2,8 @@ APPNAME =hrsc2isis
...
-	  TO=$(OUTPUT)/temp.cub \
-	  2>> $(OUTPUT)/error_message_temp.txt > /dev/null` ]; \
-	  then true; \
-	fi;
-	$(SED) 's+\[/.*/input/+\[input/+' $(OUTPUT)/error_message_temp.txt > $(OUTPUT)/error_message.txt;
-	$(RM)  $(OUTPUT)/error_message_temp.txt;
+commands:
+	$(APPNAME) FROM=$(INPUT)/h2862_0006_sr2.img \
+	           TO=$(OUTPUT)/h2862_0006_sr2.cub \
+	           > /dev/null
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/tsts/srcImage/Makefile:8]: For generic ingestion tests, I've also been adding a catlab to test the label, too in addition to the cube. Is that worth adding here?

File: isis/src/mex/apps/hrsc2isis/tsts/default/Makefile [hunk 6]
```diff
@@ -6,10 +6,4 @@ commands:
...
 	tabledump FROM=$(OUTPUT)/h0279_0000_re2.cub TO=$(OUTPUT)/h0279_0000_re2.txt NAME=LineScanTimes > /dev/null;
 	catlab FROM=$(OUTPUT)/h0279_0000_re2.cub to=$(OUTPUT)/h0279_0000_re2.pvl > /dev/null;
-	if [ `$(APPNAME) FROM=$(INPUT)/h1580_0008_sr3.img \
-	  TO=$(OUTPUT)/junk.cub \
-	   >& $(OUTPUT)/error_message_temp.txt` ]; \
-	then true; \
-	fi;
-	$(SED) 's+\[/.*/input/+\[input/+' $(OUTPUT)/error_message_temp.txt > $(OUTPUT)/error.txt;
-	$(RM)  $(OUTPUT)/error_message_temp.txt;
+
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/tsts/default/Makefile:9]: Why remove this? Was this testing a failure to ingest a SRC image?

File: isis/src/mex/apps/hrsc2isis/main.cpp [hunk 7]
```diff
@@ -11,93 +11,154 @@
...
+    msg += "] does not appear to be a Mars Express stereo or SRC file. Label keyword [DETECTOR_ID = ";
+    msg += label["DETECTOR_ID"][0] + "] is not recognized.";
     throw IException(IException::User, msg, _FILEINFO_);
   }
 
-  ProcessImportPds p;
-  Pvl label;
-  lineInFile.clear();
-  numLinesSkipped = 0;
+  // This progream is setup to work with Mex HRSC processing level 1 and 2 only.
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/main.cpp]: progream -> program

File: isis/src/mex/apps/hrsc2isis/main.cpp [hunk 8]
```diff
@@ -11,93 +11,154 @@
...
+  outCube = p.SetOutputCube("TO");
+  p.StartProcess();
+
+  Pvl otherLabels;
+  TranslateHrscLabels(label, otherLabels);
+
+  std::cout << "========================================" <<
+    otherLabels << "=======================================================" << std::endl;
+
+  if (otherLabels.hasGroup("Instrument") &&
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/main.cpp:97]: Any reason to not just write out the whole translated label at once?

File: isis/src/mex/apps/hrsc2isis/main.cpp [hunk 9]
```diff
@@ -11,93 +11,154 @@
...
+ * This function reads Mars Express HRSC Stereo files
+ *
+ * First pass through the file is a Process called with the IgnoreData() function in order to
+ * get the import class to collect prefix data. Once the prefix data is populated, we look for
+ * "gaps" - HRSC files with prefix data give us a time and exposure duration for each line, we
+ * look for where the time + exposure duration != next line's time. We populate a table
+ * (LineScanTimes) with the prefix data and lineInFile with whether or   not a gap should be
+ * inserted.                                                            
+ *                                                                      
+ * Second pass through the file is a Process called with the WriteOutp  ut function. If there were
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/main.cpp]: Kind of weird additional spaces in this paragraph: `WriteOutp ut` for example

File: isis/src/mex/apps/hrsc2isis/main.cpp [hunk 10]
```diff
@@ -172,30 +234,24 @@ void IsisMain() {
   // Get as many of the other labels as we can
   Pvl otherLabels;
 
-  //p.TranslatePdsLabels (otherLabels);
   TranslateHrscLabels(label, otherLabels);
 
-  if(otherLabels.hasGroup("Mapping") &&
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/main.cpp:178]: I think I understand why this was removed, given that mapping explicitly isn't supported. Was it supported before?

**Review by krlberry — COMMENTED** [review]
Just some comments and questions. I also found some typoes. Not even sure if this was feedback you were looking for at this point.

File: isis/src/mex/apps/hrsc2isis/tsts/srcImage/Makefile (hunk 5 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/tsts/srcImage/Makefile:8]: Yes, even if we go to a gtest

File: isis/src/mex/apps/hrsc2isis/tsts/default/Makefile (hunk 6 above)
Comment (krlberry) [reply isis/src/mex/apps/hrsc2isis/tsts/default/Makefile:9]: Resolved: yes, it was an SRC image

File: isis/src/mex/apps/hrsc2isis/main.cpp (hunk 8 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/main.cpp:97]: This was the way the previous version did it, so just didn't change it

File: isis/src/mex/apps/hrsc2isis/main.cpp (hunk 10 above)
Comment (krlberry) [reply isis/src/mex/apps/hrsc2isis/main.cpp:178]: Resolved: it was never supported.

File: isis/src/mex/apps/hrsc2isis/MexHrscBandBin.trn (hunk 1 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/MexHrscBandBin.trn]: fixed

File: isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn (hunk 2 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn]: fixed

File: isis/src/mex/apps/hrsc2isis/hrsc2isis.xml (hunk 3 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/hrsc2isis.xml]: Yes, additional description has been added

File: isis/src/mex/apps/hrsc2isis/hrsc2isis.xml (hunk 4 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/hrsc2isis.xml]: fixed

File: isis/src/mex/apps/hrsc2isis/tsts/default/Makefile (hunk 6 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/tsts/default/Makefile:9]: yes, it no longer needs to test for SRC files and fail

File: isis/src/mex/apps/hrsc2isis/main.cpp (hunk 7 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/main.cpp]: fixed

File: isis/src/mex/apps/hrsc2isis/main.cpp (hunk 9 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/main.cpp]: fixed

File: isis/src/mex/apps/hrsc2isis/main.cpp (hunk 10 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/main.cpp:178]: No it was not. From what I can tell this code was copied directly from pds2isis. The mapping section was just not deleted.

File: isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn [hunk 11]
```diff
@@ -88,6 +79,15 @@ Group = TargetName
   Translation    = (Cal, CAL)
 End_Group
 
+Group = Summing
+  Auto
+  InputKey       = MACROPIXEL_SIZE
+  Default        = 1
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn]: Needs to be updated to `InputDefault`.

File: isis/src/mex/apps/hrsc2isis/MexHrscKernels.trn [hunk 12]
```diff
@@ -0,0 +1,21 @@
+# Translations for HRSC Stereo and SRC instrument.
+# ISIS cube label Kernels group
+
+Group = NaifIkCode
+  Auto
+  InputKey       = DETECTOR_ID
+  OutputName     = DetectorId
+  OutputPosition = (Group, Archive)
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/MexHrscKernels.trn]: Archive -> Kernels.

**Review by krlberry — CHANGES_REQUESTED** [review]

File: isis/src/mex/apps/hrsc2isis/MexHrscKernels.trn (hunk 12 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/MexHrscKernels.trn]: thanks

File: isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn (hunk 11 above)
Comment (scsides) [reply isis/src/mex/apps/hrsc2isis/MexHrscInstrument.trn]: thanks

**Comment by jlaura** [comment]
[SRC ingestion app](https://trello.com/c/N7zPzcYU/186-src-ingestion-app)

File: isis/src/mex/apps/hrsc2isis/hrsc2isis.xml [hunk 13]
```diff
@@ -5,18 +5,26 @@
...
-    not a gap should be inserted. For all files, we now process the  data and if 
-    there were gaps, they will be put in their proper places. This is a two-pass
-    system for files with prefix data, one-pass for files without.  The Isis2
-    equivalent to this program is mex2isis.pl. It is worth noting that
-    regardless of the  input file's byte order, the prefix data byte order is 
-    always LSB.
+    <p>
+      This program imports Mars Express HRSC files. This includes both the HRSC 
+      Stereo and the SRC instruments with procesing level of "1" or "2" (not "3"). 
+      The program populates all label keywords necessary to create the corrisponding camera
```
Comment (krlberry) [inline isis/src/mex/apps/hrsc2isis/hrsc2isis.xml]: corrisponding -> corresponding

**Review by krlberry — DISMISSED** [review]
:+1: I found a single typo in the docs, otherwise looks great!

**Review by krlberry — APPROVED** [review]

**Comment by jlaura** [comment]
@kberryUSGS @scsides Is this good to merge? I see one approval, but not sure if you all would like another set of eyes?

**Pull request merged**